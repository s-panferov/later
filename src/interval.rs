use chrono::{DateTime, Duration, Utc};
use std::fmt::Debug;

#[derive(Debug, Clone)]
pub struct Interval {
	pub from: DateTime<Utc>,
	pub to: Option<DateTime<Utc>>,
}

impl Interval {
	pub fn from_now() -> Self {
		Interval {
			from: Utc::now(),
			to: None,
		}
	}

	pub fn from(from: DateTime<Utc>) -> Self {
		Interval { from, to: None }
	}
}

/// Something that resolves to the range of time
pub trait AsIntervals: Debug {
	/// Get the estimate of the interval
	fn duration_hint(&self) -> Duration;

	/// Get an interator to resolve intervals
	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>>;
}
