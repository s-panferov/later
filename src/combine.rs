use crate::interval::Interval;
use std::cmp::Ordering;

#[derive(Clone, Debug)]
enum Variant {
	Finished(usize),
	Fresh(usize, Interval),
	Taken(usize),
}

type Iterators = Vec<Box<dyn Iterator<Item = Interval>>>;

/// An iterator that combines a set of other iterators
/// and produces a consequest stream of Intervals without overlaps.
struct Combine {
	iterators: Iterators,
	state: Vec<Variant>,
	prev: Option<Interval>,
	completed: bool,
}

impl Combine {
	fn new(iterators: Iterators) -> Self {
		let state = iterators
			.iter()
			.enumerate()
			.map(|(i, _)| Variant::Taken(i))
			.collect();

		Combine {
			iterators,
			state,
			prev: None,
			completed: false,
		}
	}
}

impl Iterator for Combine {
	type Item = Interval;

	fn next(&mut self) -> Option<Interval> {
		if self.completed {
			return None;
		}

		loop {
			let mut indexes: Vec<(usize, usize)> = vec![];

			self.state.iter().enumerate().for_each(|(i, v)| match v {
				Variant::Finished(_) => {}
				Variant::Fresh(_, _) => {}
				Variant::Taken(j) => indexes.push((i, j.clone())),
			});

			for (i, j) in indexes {
				let iterator = self.iterators.get_mut(j).unwrap();
				self.state[i] = match iterator.next() {
					Some(interval) => Variant::Fresh(j, interval),
					None => Variant::Finished(j),
				}
			}

			self.state.sort_by(|a, b| match a {
				Variant::Finished(_) => Ordering::Less,
				Variant::Taken(_) => Ordering::Less,
				Variant::Fresh(_, a) => match b {
					Variant::Finished(_) => Ordering::Greater,
					Variant::Taken(_) => Ordering::Greater,
					Variant::Fresh(_, b) => {
						if a.from > b.from {
							Ordering::Greater
						} else if a.from < b.from {
							Ordering::Less
						} else {
							if a.to.is_none() {
								Ordering::Greater
							} else if b.to.is_none() {
								Ordering::Less
							} else {
								a.to.unwrap().cmp(&b.to.unwrap())
							}
						}
					}
				},
			});

			println!("{:?}", self.state);

			let mut result: Option<Interval> = None;
			for v in &mut self.state {
				match v {
					Variant::Finished(_) => {}
					Variant::Taken(_) => {}
					Variant::Fresh(i, _) => {
						let mut tmp = Variant::Taken(i.clone());
						std::mem::swap(v, &mut tmp);
						match tmp {
							Variant::Finished(_) => {}
							Variant::Taken(_) => {}
							Variant::Fresh(_, v) => result = Some(v),
						}
						break;
					}
				}
			}

			if result.is_some() {
				let interval = result.as_mut().unwrap();
				if interval.to.is_none() {
					// If our current interval does not have an end we don't need to produce
					// anything else because it would overlap.
					self.completed = true;
					return result;
				} else {
					if self.prev.is_some() {
						let to = self.prev.as_ref().unwrap().to.unwrap();
						if interval.from < to {
							interval.from = to.clone();
							if interval.from > interval.to.unwrap() {
								// We consumed the whole interval
								continue;
							}
						}
					}
					self.prev = result.clone();
					return result;
				}
			} else {
				self.completed = true;
				return None;
			}
		}
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::interval::AsIntervals;
	use chrono::Weekday;
	use insta::assert_debug_snapshot_matches;

	#[test]
	fn combine_weekend() {
		let stream = Combine::new(vec![
			Weekday::Sat.iter_within(Interval::from(
				"2019-01-01T00:00:00Z".parse().unwrap(),
			)),
			Weekday::Sun.iter_within(Interval::from(
				"2019-01-01T00:00:00Z".parse().unwrap(),
			)),
		]);

		let what: Vec<Interval> = stream.take(10).collect();
		assert_debug_snapshot_matches!("combine_weekend", what);
	}
}
