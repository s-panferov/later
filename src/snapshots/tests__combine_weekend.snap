---
created: "2019-08-31T23:26:51.551620Z"
creator: insta@0.10.1
source: src/combine.rs
expression: what
---
[
    Interval {
        from: 2019-01-05T00:00:00Z,
        to: Some(
            2019-01-05T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-01-06T00:00:00Z,
        to: Some(
            2019-01-06T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-01-12T00:00:00Z,
        to: Some(
            2019-01-12T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-01-13T00:00:00Z,
        to: Some(
            2019-01-13T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-01-19T00:00:00Z,
        to: Some(
            2019-01-19T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-01-20T00:00:00Z,
        to: Some(
            2019-01-20T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-01-26T00:00:00Z,
        to: Some(
            2019-01-26T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-01-27T00:00:00Z,
        to: Some(
            2019-01-27T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-02-02T00:00:00Z,
        to: Some(
            2019-02-02T23:59:59.999999999Z,
        ),
    },
    Interval {
        from: 2019-02-03T00:00:00Z,
        to: Some(
            2019-02-03T23:59:59.999999999Z,
        ),
    },
]
